[MRG] Libmbin/relative offset calc within a class

monkeyman192 opened the pull request:
This adds 3 new functions (2 static and one not) to `NMSTemplate` which gets the offset of a given field within a class. This can be used when writing programs which find stuff in memory, as it will allow offsets of fields to be determined automatically, instead of having to be hard-coded

monkeyman192 committed:
Add new OffsetOf function to NMSTemplate to get the offset of a field within a class

cmkushnir commented:
Like OffsetOf, the standard function names (C# and C++) for size and alignment are SizeOf, and AlignOf.
May want to add an int INDEX param to the OffsetOf methods:
public static int OffsetOf(Type type, string fieldName, int index = 0) {
...
if (fieldName == field.Name) {
    if( index > 0 ) offset += adjustment to get specific element in array or List, may also want validate that type is array or List
    return offset;
}
...

chrispurnell commented:
If GetSize() fails you have it returning 0.  It might be better to throw an exception.

chrispurnell commented on libMBIN/Source/Template/NMSTemplate.cs:
```diff
@@ -103,13 +103,117 @@ public class NMSTemplate
+                int alignment = GetAlignment(field.FieldType);
+                // Make sure the alignment is taken into consideration.
+                if (offset % alignment != 0) {
+                    offset += (offset % alignment);
```
Shouldn't this be:
offset += alignment - (offset % alignment);

monkeyman192 committed:
Fix some bugs and standardise names of functions

monkeyman192 commented:
@chrispurnell and @cmkushnir I have (finally!) updated this PR!
I have taken in feedback from both of your comments and updated accordingly.
Turns out there were some more issues than I realised which were uncovered when I wrote a simple unit test (also, I have updated the unit test dependencies so maybe these can be fixed too at some point...)
Hopefully this is now good to go 😄

cmkushnir commented:
Took a quick look, nothing jumped out.  Would suggest making NMSTemplateMap public, other than that looks good.  👍🏻

monkeyman192 commented:
> 
> 
> Took a quick look, nothing jumped out. Would suggest making NMSTemplateMap public, other than that looks good. 👍🏻

TBH, I might just leave it. It's not my change and in the interest of getting this merged it's probably not really needed haha. Unless you feel strongly about it. Also probably got a new NMS update dropping soon so would be good to have these PR's sorted! (I need to deal with the .net 4 one also...)

cmkushnir commented:
That's fine.  I only mentioned it because I needed it for my app, but since it wasn't public had to create my own version.

monkeyman192 commented:
> 
> 
> That's fine. I only mentioned it because I needed it for my app, but since it wasn't public had to create my own version.

Ah ok! I'm surprised you didn't just make a PR to make it public 😛
